### PR TITLE
Added request attributes to extract data from request

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes part of #274
| Related issues/PRs | replaces #2054
| License | MIT
| Documentation PR | none

#### What's in this PR?

This PR added the class request attributes and providers for that. The providers extract the attributes from the request and returns it to the request-analyzer.

#### Why?

I need it for the custom-url feature which will update the locale of a request in a different way than the normal page does it.

#### Example Usage

~~~php
<?php

class RequestAttributesProvider implements RequestAttributesProviderInterface
{
    /**
     * {@inheritdoc}
     */
    public function getAttributes(Request $request)
    {
        $attributes = [];

...

        return new RequestAttributes($attributes);
    }
}
~~~

~~~xml
<service id="sulu_core.request_attributes.admin" class="RequestAttributesProvider">
    <tag name="sulu.context" context="website"/>
    <tag name="sulu.request_attributes" priority="0"/>
</service>
~~~

With the context you can define where this provider will be used (website or admin).

### CODE REVIEW

File: src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php [hunk 1]
```diff
@@ -64,7 +64,12 @@ public function showAction(
     ) {
         $code = $exception->getStatusCode();
         $showException = $request->attributes->get('showException', $this->debug);
-        $template = $this->requestAnalyzer->getWebspace()->getTheme()->getErrorTemplate($code);
+
+        $template = null;
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/WebsiteBundle/Controller/ExceptionController.php]: this change replaces #2054

File: src/Sulu/Component/Webspace/Tests/Functional/Analyzer/RequestAnalyzerTest.php [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    private $requestAnalyzer;
+
+    /**
+     * @var PHPUnit_Framework_MockObject_MockObject
+     */
+    private $webspaceManager;
+
+    public function setUp()
+    {
+        $this->webspaceManager = $this->getMockForAbstractClass(
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Tests/Functional/Analyzer/RequestAnalyzerTest.php]: Was this test only moved? Wondering why the phpunit mocking is used here.

**Comment by danrot** [comment]
What's the reason for splitting `RequestAnalyzer` and `WebsiteRequestAnalyzer`?

File: src/Sulu/Component/Webspace/Tests/Functional/Analyzer/RequestAnalyzerTest.php (hunk 2 above)
Comment (wachterjohannes) [inline src/Sulu/Component/Webspace/Tests/Functional/Analyzer/RequestAnalyzerTest.php]: this is the same logic as before i have not changed here anything

**Comment by wachterjohannes** [comment]
because there are some additional checks in WebsiteRequestAnalyzer. They could not really be moved to the provider because then no other provider will get called when the exception is thrown.

**Comment by danrot** [comment]
Ok, I see, and you still want them to be executed only on the website right? And isn't the exception stopping execution anyway, no matter where it is thrown?

File: src/Sulu/Bundle/CoreBundle/DependencyInjection/Compiler/RequestAnalyzerCompilerPass.php [hunk 3]
```diff
@@ -22,63 +19,21 @@
...
-
-        if ($container->getParameter('sulu.context') === SuluKernel::CONTEXT_WEBSITE) {
-            $container->setAlias('sulu_core.webspace.request_analyzer', 'sulu_core.webspace.request_analyzer.website');
-        } else {
-            $container->setAlias('sulu_core.webspace.request_analyzer', 'sulu_core.webspace.request_analyzer.admin');
+        $references = [];
+        foreach ($container->findTaggedServiceIds(self::TAG_NAME) as $id => $tags) {
+            foreach ($tags as $attributes) {
+                $priority = array_key_exists('priority', $attributes) ? $attributes['priority'] : 0;
+                $references[$priority] = new Reference($id);
```
Comment (danrot) [inline src/Sulu/Bundle/CoreBundle/DependencyInjection/Compiler/RequestAnalyzerCompilerPass.php]: The priority should be used as in the event dispatcher, i.e. The priority should be reversed and each priority must be assignable more than once.

**Comment by wachterjohannes** [comment]
@danrot this should be green now.

File: src/Sulu/Component/Webspace/Analyzer/RequestAnalyzer.php [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+/**
+ * Default request analyzer will be used for sulu-admin and extended for sulu-website.
+ */
+class RequestAnalyzer implements RequestAnalyzerInterface
+{
+    /**
+     * @var RequestProcessorInterface[]
+     */
+    private $requestAttributesProvider;
```
Comment (danrot) [inline src/Sulu/Component/Webspace/Analyzer/RequestAnalyzer.php]: Should also be renamed.

**Comment by danrot** [comment]
Sry that I already merged it, you should rename the variable I have mentioned here. And can you maybe create an issue for completely replacing the `RequestAnalyzer` ( and just keeping it for BC).

**Comment by wachterjohannes** [comment]
https://github.com/sulu-io/sulu/issues/2135 and https://github.com/sulu-io/sulu/commit/89aba8b3f2713996cb6b871a192a2c59d3c4cca3

**Pull request merged**